### PR TITLE
First DNA with UI RAD

### PR DESCRIPTION
**Pull request by JettTech**
#### `hc:generate-dna` will:
 - create a dna-src folder with all the dna boilerplate
 - create a zome for every zome field and lib file (optional field in type-spec)
 - create an entry folder for every entry in each zome
 - generate a mod, handlers, and validation file according to the type-spec fields for each entry
 - generate a test folder and index file
 - generate tests for each entry

Changes to note:
  - optional updates to the type-spec
  - directory updates and name changes
  - switched to npm from yarn (for testing / ci)

---
#### To generate DNA only:
 - clone repo
 - enter nix-shell
 - run npm `hc:generate-dna <type-spec.json>` 
     >NB: *The type-spec.json is optional and will default to th sample provided in the setup/type-specs folder.*

##### To run tests:
-  run `npm i`
-  run `npm hc:test`

#### To generate full HC HApp:
 - clone repo
 - enter nix-shell
 - run npm `happ:generate`

##### To run happ:
-  run `npm i`
-  run `npm start`

### CODE REVIEW

File: conductor-config.example.toml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+exclude = false
+pattern = "^err/"
+
+[[logger.rules.rules]]
+color = "white"
+exclude = false
+pattern = "^debug/dna"
+
+[network]
+sim2h_url = "ws://public.sim2h.net:9000"
```
Comment (zo-el) [inline conductor-config.example.toml]: Please use localhost here

**Review by zo-el — CHANGES_REQUESTED** [review]
Not reviewed everything, this is just the first run of reviewing.
I think we need to rename `setup` to `src` and remove `dna-src`. it should be gitignored right?

Also we need to give a link or notes of how to create a `sample-type-spec.json` file.

**Comment by JettTech** [comment]
Yes, they should be ignored.  Sorry for the massive file overload.  I removed the files and add ui-src and dna-src to the .gitignore.

**Review by zo-el — CHANGES_REQUESTED** [review]
### Review: (These are just overview questions and thoughts)
- Why is the agent-setup and type-spec still in set up? it should all be inside the src. All the code should be inside src
- Read me needs to be updated
- Cool Hint to notify that you may not be running in nix-shell. Later we could probably change the color of the Hint to Yellow or something
- Why don't we implement a way to refer you own path?
- So we are only generating one zome right?
-  Can we write some comments later on the code. Not everywhere but in places like when your reading the sample-type-spec, it would help the reader() not very important.

### Code Review: 
- **BUG needs attention:** The `default.nix` file that is generated by `hc init` (I am assuming) points to a very old release of holochain. (` holonix-release-tag = "further-bad-assigning";`) this needs to be updated

**Comment by JettTech** [comment]
@joel 
 - Why is the agent-setup and type-spec still in setup? it should all be inside the src. All the code should be inside src/
    - The setup folder will remain in the user's own app directory when using the holonix tool for optional continued use, while the code generation src will be removed.

> NB: I decided to keep it inside a separate setup directory, but place that directory as a sub-directory in the src folder. Open to thoughts on this.

 - Read me needs to be updated 
    -  :+1: This is a next todo.

 - Cool Hint to notify that you may not be running in nix-shell. Later we could probably change the color of the Hint to Yellow or something
    - Good idea.  I decided to import and use Chalk. 

 - Why don't we implement a way to refer your own path?
    -  Agreed and implemented.

 - So we are only generating one zome right?
   - The code can generate multiple, but only one is required.

 - Can we write some comments later on the code. Not everywhere but in places like when you're reading the sample-type-spec, it would help the reader() not very important of course
   - Can you name the places that stood out to you? :question:

File: package.json [hunk 2]
```diff
@@ -1,23 +1,54 @@
...
+    "ui:generate": "node src/ui-setup/scripts/generate-ui",
+    "ui:generate-clean": "rm -rf ui-src/*",
+    "hc:clean-storage": "rm -rf ./.holochain/holo/storage",
+    "hc:generate-dna": "node src/dna-setup/scripts/generate-dna",
+    "hc:config-add-agent": "node setup/conductor-setup/add-agent-to-conductor",
+    "hc:keygen": "node setup/conductor-setup/create-new-agent",
+    "hc:package": "cd dna-src && hc package",
+    "hc:build": "npm run hc:package",
+    "hc:test": "cd dna-src && RUST_BACKTRACE=1 HC_IGNORE_SIM2H_URL_PROPERTY=true hc test | node test/node_modules/faucet/bin/cmd.js",
+    "hc:start": "holochain -c conductor-config.toml &> conductor.log &",
```
Comment (robbiecarlton) [inline package.json]: It's a little awkward that this hides its output. It means it fails silently when you haven't generated a conductor. We could add a check to this command that returns an error if there's no conductor, saying "try running hc:generate-conductor first".

But actually I think it's better to not send the output to a log file. Let the user do that if they want.

File: src/setup/type-specs/sample-type-spec.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "types": {
```
Comment (robbiecarlton) [inline src/setup/type-specs/sample-type-spec.json:2]: this file should have two types defined in it. Having two types in this one zome doesn't currently work.

**Review by robbiecarlton — CHANGES_REQUESTED** [review]
Adding a second type to a single zome is currently broken. I get an error
```
Holochain Instance Error: Zome function 'create_post' not found in Zome 'users'
```
when trying to call `create` on my second type.

I'm also getting a validation error on my first type but that might be user error.

File: README.md [hunk 4]
```diff
@@ -1,15 +1,114 @@
...
+    ```
+    $ npm run happ:generate <type-spec.json>
+    ```
+
+
+### Run your custom happ:
+
+1.  Generate your Holochain Conductor
+    ```
+    $ npm run hc:generate-conductor
```
Comment (robbiecarlton) [inline README.md]: remove this step

File: README.md [hunk 5]
```diff
@@ -1,15 +1,114 @@
...
+    $ nix-shell
+    ```
+3. Install root node dependencies
+    ```
+    $ npm i
+    ```
+
+4. Generate your custom Holochain Happ according to the type-spec.json file.
+    >NB: *The type-spec.json is optional and will default to th sample provided in the src/setup/type-specs folder.*
+    ```
```
Comment (robbiecarlton) [inline README.md]: add a brief description/example of the json

File: README.md [hunk 6]
```diff
@@ -1,15 +1,114 @@
...
+$ nix-shell
+```
+## 2 - Run the hc-happ-scaffold command
+From within the `nix-shell` environment, first create a directory for all your Holochain projects (if you haven't already). You can create it wherever you like; here's a recommended setup:
+
 ```
-$ yarn generate:ui sample-type-spec.json
+$ cd ~
+$ mkdir Holochain
+$ cd Holochain
```
Comment (robbiecarlton) [inline README.md]: move this to a Readme in Holonix

File: src/dna-setup/scripts/generate-dna-zomes.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  else return stdout.trim()
+}
+
+async function findTestDirPath (zomeDir) {
+  const { stderr, stdout } = await exec(`cd ${zomeDir} && cd ../../../../test && echo $(pwd -P)`)
+  if(stderr) console.error('stderr:', stderr)      
+  else return testDirectoryPath = stdout.trim()
+}
+
+const renderZomeCargoToml = (zomeName, zomeDir) => {
```
Comment (robbiecarlton) [inline src/dna-setup/scripts/generate-dna-zomes.js:28]: if it writes the file, we call it `generate`. If it just returns the content to be written, we call it `render`

File: src/dna-setup/scripts/generate-dna-zomes.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  const pathAsArray = zomeDir.split('/')
+  const zomeCodeDir = pathAsArray.splice(0, pathAsArray.length-2).join('/')
+  const cargoTomlTemplatePath = path.resolve("src/dna-setup/zome-template", "Cargo.toml");
+  const cargoTomlTemplate = fs.readFileSync(cargoTomlTemplatePath, 'utf8')
+  const cargoToml = replaceNamePlaceHolders(cargoTomlTemplate, ZOME_NAME, zomeName)
+  const writeCargoToml = fs.writeFileSync(`${zomeCodeDir}/Cargo.toml`, cargoToml)
+  return writeCargoToml
+}
+
+const renderNixSetup = (dnaName, zomeDir) => {
```
Comment (robbiecarlton) [inline src/dna-setup/scripts/generate-dna-zomes.js:38]: `generateNixSetup`

File: src/setup/type-specs/sample-type-spec.json [hunk 9]
```diff
@@ -6,6 +6,13 @@
       "definition": {
         "name": "string",
         "avatar_url": "string"
+      },
+      "functions": {
+        "create": true,
+        "update": true,
+        "delete": true,
+        "get": true,
+        "list": true
```
Comment (robbiecarlton) [inline src/setup/type-specs/sample-type-spec.json:15]: do we currently support this?

**Review by robbiecarlton — APPROVED** [review]

File: src/setup/type-specs/sample-type-spec.json (hunk 9 above)
Comment (JettTech) [reply src/setup/type-specs/sample-type-spec.json:15]: The DNA does - I'll move it to that section of the readme.the DNA does - I'll move it to that section of the readme.

File: src/dna-setup/scripts/generate-dna-zomes.js (hunk 8 above)
Comment (JettTech) [reply src/dna-setup/scripts/generate-dna-zomes.js:38]: I like the naming distinction.

**Pull request merged**